lib/widgets/compose_box.dart

```diff
@@ -891,43 +886,62 @@ class _SendButtonState extends State<_SendButton> {
   @override
   void initState() {
     super.initState();
-    widget.topicController?.hasValidationErrors.addListener(_hasErrorsChanged);
-    widget.contentController.hasValidationErrors.addListener(_hasErrorsChanged);
+    final controller = widget.controller;
+    if (controller is StreamComposeBoxController) {
+      controller.topic.hasValidationErrors.addListener(_hasErrorsChanged);
+    }
+    controller.content.hasValidationErrors.addListener(_hasErrorsChanged);
   }
 
   @override
   void didUpdateWidget(covariant _SendButton oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (widget.topicController != oldWidget.topicController) {
-      oldWidget.topicController?.hasValidationErrors.removeListener(_hasErrorsChanged);
-      widget.topicController?.hasValidationErrors.addListener(_hasErrorsChanged);
+
+    final controller = widget.controller;
+    final oldController = oldWidget.controller;
+    if (controller == oldController) return;
+
+    if (oldController is StreamComposeBoxController) {
+      oldController.topic.hasValidationErrors.removeListener(_hasErrorsChanged);
     }
-    if (widget.contentController != oldWidget.contentController) {
-      oldWidget.contentController.hasValidationErrors.removeListener(_hasErrorsChanged);
-      widget.contentController.hasValidationErrors.addListener(_hasErrorsChanged);
+    if (controller is StreamComposeBoxController) {
+      controller.topic.hasValidationErrors.addListener(_hasErrorsChanged);
     }
+    oldController.content.hasValidationErrors.removeListener(_hasErrorsChanged);
+    controller.content.hasValidationErrors.addListener(_hasErrorsChanged);
   }
 
   @override
   void dispose() {
-    widget.topicController?.hasValidationErrors.removeListener(_hasErrorsChanged);
-    widget.contentController.hasValidationErrors.removeListener(_hasErrorsChanged);
+    final controller = widget.controller;
+    if (controller is StreamComposeBoxController) {
+      controller.topic.hasValidationErrors.removeListener(_hasErrorsChanged);
+    }
+    controller.content.hasValidationErrors.removeListener(_hasErrorsChanged);
     super.dispose();
   }
 
   bool get _hasValidationErrors {
-    return (widget.topicController?.hasValidationErrors.value ?? false)
-      || widget.contentController.hasValidationErrors.value;
+    bool result = false;
+    final controller = widget.controller;
+    if (controller is StreamComposeBoxController) {
+      result = controller.topic.hasValidationErrors.value;
+    }
+    result |= controller.content.hasValidationErrors.value;
+    return result;
```

Comment:
I feel that this is an indicator to further abstract away logic around validation errors, probably as a future follow-up.

Reply:
I'm not yet picturing what that might look like, so I won't add a TODO in this revision, but I'm happy to if you help me understand better (or you can push that TODO yourself, wiki-style, after this is merged).